File: scripts/release/check_versions_win.py
import pathlib, re, sys, subprocess, json

fail = 0

# 1) Get Cargo.toml workspace version
root_cargo = pathlib.Path("Cargo.toml")
if not root_cargo.exists():
    print("error: Cargo.toml not found")
    sys.exit(1)
text = root_cargo.read_text(encoding='utf-8')
m = re.search(r'^version = "([^"]+)"', text, re.MULTILINE)
if not m:
    print("error: could not find version in Cargo.toml")
    sys.exit(1)
workspace_version = m.group(1)

# 2) Get npm version
npm_pkg = pathlib.Path("npm/helpofai/package.json")
if not npm_pkg.exists():
    print("error: npm/helpofai/package.json not found")
    sys.exit(1)
try:
    pkg = json.loads(npm_pkg.read_text(encoding='utf-8'))
    npm_version = pkg.get("version")
except Exception as e:
    print(f"error parsing package.json: {e}")
    sys.exit(1)

if workspace_version != npm_version:
    print(f"error: npm version ({npm_version}) does not match workspace version ({workspace_version})")
    fail = 1

# 3) Get facts version
facts_file = pathlib.Path("web/lib/facts.generated.ts")
if not facts_file.exists():
    print("error: web/lib/facts.generated.ts not found")
    sys.exit(1)
facts_text = facts_file.read_text(encoding='utf-8')
m_facts = re.search(r'"version":\s*"([^"]+)"', facts_text)
if not m_facts:
    print("error: could not find version in facts.generated.ts")
    sys.exit(1)
facts_version = m_facts.group(1)

if workspace_version != facts_version:
    print(f"error: facts version ({facts_version}) does not match workspace version ({workspace_version})")
    fail = 1

# 4) Verify Cargo.lock sync
try:
    res = subprocess.run(["cargo", "metadata", "--locked", "--format-version", "1", "--no-deps"], capture_output=True, text=True)
    if res.returncode != 0:
        print("error: Cargo.lock is out of sync with manifests.")
        fail = 1
except Exception as e:
    print(f"error running cargo metadata: {e}")
    fail = 1

if fail == 0:
    print(f"Version state OK: workspace={workspace_version}, npm={npm_version}, facts={facts_version}, lockfile in sync.")
else:
    print("Version check failed.")
sys.exit(fail)

File: scripts/release/check_versions_win.js
const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

let fail = 0;

// 1) Get Cargo.toml workspace version
const rootCargo = 'Cargo.toml';
if (!fs.existsSync(rootCargo)) {
    console.error('error: Cargo.toml not found');
    process.exit(1);
}
const cargoText = fs.readFileSync(rootCargo, 'utf8');
const m = cargoText.match(/^version = "([^"]+)"/m);
if (!m) {
    console.error('error: could not find version in Cargo.toml');
    process.exit(1);
}
const workspaceVersion = m[1];

// 2) Get npm version
const npmPkgPath = 'npm/helpofai/package.json';
if (!fs.existsSync(npmPkgPath)) {
    console.error('error: npm/helpofai/package.json not found');
    process.exit(1);
}
let npmVersion;
try {
    const pkg = JSON.parse(fs.readFileSync(npmPkgPath, 'utf8'));
    npmVersion = pkg.version;
} catch (e) {
    console.error(`error parsing package.json: ${e}`);
    process.exit(1);
}

if (workspaceVersion !== npmVersion) {
    console.error(`error: npm version (${npmVersion}) does not match workspace version (${workspaceVersion})`);
    fail = 1;
}

// 3) Get facts version
const factsFile = 'web/lib/facts.generated.ts';
if (!fs.existsSync(factsFile)) {
    console.error('error: web/lib/facts.generated.ts not found');
    process.exit(1);
}
const factsText = fs.readFileSync(factsFile, 'utf8');
const mFacts = factsText.match(/"version":\s*"([^"]+)"/);
if (!mFacts) {
    console.error('error: could not find version in facts.generated.ts');
    process.exit(1);
}
const factsVersion = mFacts[1];

if (workspaceVersion !== factsVersion) {
    console.error(`error: facts version (${factsVersion}) does not match workspace version (${workspaceVersion})`);
    fail = 1;
}

// 4) Verify Cargo.lock sync
try {
    execSync('cargo metadata --locked --format-version 1 --no-deps', { stdio: 'ignore' });
} catch (e) {
    console.error('error: Cargo.lock is out of sync with manifests.');
    fail = 1;
}

if (fail === 0) {
    console.log(`Version state OK: workspace=${workspaceVersion}, npm=${npmVersion}, facts=${factsVersion}, lockfile in sync.`);
} else {
    console.error('Version check failed.');
}

process.exit(fail);
